comp/collector/collector/collectorimpl/agent_check_metadata.go

```diff
@@ -0,0 +1,140 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+package collectorimpl
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/DataDog/datadog-agent/comp/core/autodiscovery/autodiscoveryimpl"
+	hostMetadataUtils "github.com/DataDog/datadog-agent/comp/metadata/host/hostimpl/utils"
+	"github.com/DataDog/datadog-agent/pkg/collector"
+	"github.com/DataDog/datadog-agent/pkg/collector/externalhost"
+	"github.com/DataDog/datadog-agent/pkg/collector/runner/expvars"
+	"github.com/DataDog/datadog-agent/pkg/serializer/marshaler"
+	jmxStatus "github.com/DataDog/datadog-agent/pkg/status/jmx"
+	"github.com/DataDog/datadog-agent/pkg/util/hostname"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+//
+// The "agent_check" metadata payload contains information about all running checks and the additional hostnames they
+// added to the Agent.
+//
+
+const (
+	defaultInterval   = 10 * time.Minute
+	firstPayloadDelay = 1 * time.Minute
+)
+
+// Payload handles the JSON unmarshalling of the metadata payload
+type Payload struct {
+	hostMetadataUtils.CommonPayload
+	Meta             hostMetadataUtils.Meta `json:"meta"`
+	AgentChecks      []interface{}          `json:"agent_checks"`
+	ExternalhostTags externalhost.Payload   `json:"external_host_tags"`
+}
+
+// MarshalJSON serialization a Payload to JSON
+func (p *Payload) MarshalJSON() ([]byte, error) {
+	// use an alias to avoid infinite recursion while serializing
+	type PayloadAlias Payload
+
+	return json.Marshal((*PayloadAlias)(p))
+}
+
+// SplitPayload breaks the payload into times number of pieces
+func (p *Payload) SplitPayload(_ int) ([]marshaler.AbstractMarshaler, error) {
+	return nil, fmt.Errorf("AgentChecks Payload splitting is not implemented")
+}
+
+// GetPayload builds a payload of all the agentchecks metadata
+func (c *collectorImpl) GetPayload(ctx context.Context) *Payload {
+	hostname, _ := hostname.Get(ctx)
+
+	meta := hostMetadataUtils.GetMetaFromCache(ctx, c.config)
+	meta.Hostname = hostname
+
+	cp := hostMetadataUtils.GetCommonPayload(hostname, c.config)
+	payload := &Payload{
+		CommonPayload:    *cp,
+		Meta:             *meta,
+		ExternalhostTags: *externalhost.GetPayload(),
+	}
+
+	checkStats := expvars.GetCheckStats()
+	for _, stats := range checkStats {
+		for _, s := range stats {
+			var status []interface{}
+			if s.LastError != "" {
+				status = []interface{}{
+					s.CheckName, s.CheckName, s.CheckID, "ERROR", s.LastError, "",
+				}
+			} else if len(s.LastWarnings) != 0 {
+				status = []interface{}{
+					s.CheckName, s.CheckName, s.CheckID, "WARNING", s.LastWarnings, "",
+				}
+			} else {
+				status = []interface{}{
+					s.CheckName, s.CheckName, s.CheckID, "OK", "", "",
+				}
+			}
+			if status != nil {
```

Comment:
If there a case in which status is nil? Mostly asking as there is the `else` branch which will always get reach of the previous two are not met. Maybe we can remove the `status != nil`

Reply:
I think we can remove it. I move this code from the old check. Not sure why git can't track the change.